kroxylicious/src/main/java/io/kroxylicious/proxy/internal/codec/KafkaRequestDecoder.java

```diff
@@ -89,29 +90,83 @@ protected Frame decodeHeaderAndBody(ChannelHandlerContext ctx, ByteBuf in, final
                 log().trace("{}: body {}", ctx, body);
             }
 
-            frame = new DecodedRequestFrame<ApiMessage>(apiVersion, correlationId, decodeResponse, header, body);
+            frame = new DecodedRequestFrame<>(apiVersion, correlationId, decodeResponse, header, body);
             if (log().isTraceEnabled()) {
                 log().trace("{}: frame {}", ctx, frame);
             }
         }
         else {
+            boolean hasResponse = true;
+            if (apiKey == ApiKeys.PRODUCE) {
+                short acks = readAcks(in, startOfMessage, apiKey.id, apiVersion);
+                hasResponse = acks != 0;
+            }
             in.readerIndex(sof);
-            frame = opaqueFrame(in, correlationId, decodeResponse, length);
+            frame = opaqueFrame(in, correlationId, decodeResponse, length, hasResponse);
             in.readerIndex(sof + length);
         }
 
         return frame;
     }
 
+    private static void incrementReaderIndex(ByteBuf byteBuf, int increment) {
+        byteBuf.readerIndex(byteBuf.readerIndex() + increment);
+    }
+
+    static short readAcks(ByteBuf in, int startOfMessage, short apiKey, short apiVersion) {
+        // Annoying case: we need to know whether to expect a response so that we know
+        // whether to add to the correlation (so that, in turn, we know how to rewrite the correlation
+        // id of the client response).
+        // Adding ack-less Produce requests to the correlation => OOME.
+        // This requires decoding at least the first one or two
+        // fields of all Produce requests.
+        // Because we want to avoid parsing the produce request using ProduceRequestData
+        // just for this we are stuck with hand coding deserialization code...
+        in.readerIndex(startOfMessage);
+        short headerVersion = ApiKeys.forId(apiKey).requestHeaderVersion(apiVersion);
+        incrementReaderIndex(in, 4);
+        if (headerVersion >= 1) {
+            int clientIdLength = in.readShort();
+            incrementReaderIndex(in, clientIdLength);
+        }
+        if (headerVersion >= 2) {
+            int numTaggedFields = ByteBufAccessorImpl.readUnsignedVarint(in);
+            for (int i = 0; i < numTaggedFields; i++) {
+                ByteBufAccessorImpl.readUnsignedVarint(in);
+                int size = ByteBufAccessorImpl.readUnsignedVarint(in);
+                incrementReaderIndex(in, size);
+            }
+        }
+
+        final short acks;
+        if (apiVersion >= 3) { // Transactional id comes before acks
+            int transactionIdLength;
+            if (apiVersion < 9) { // Last non-flexible version
+                transactionIdLength = in.readShort();
+            }
+            else if (apiVersion == 9) { // First flexible version
+                transactionIdLength = ByteBufAccessorImpl.readUnsignedVarint(in);
+            }
+            else {
```

Comment:
I think forcing a breaking change is okay here.  If Kafka were to evolve ProduceRequest in an incompatible way, existing Kroxylicious deployments wouldn't try to use that version (owing to ApiVersion negotiation).  So it wouldn't break existing deployments even if they are pointed at newer Kafka Brokers.

There would have to be a new version of Kroxylicious, and this approach would force us to examine the schema.